File: widgets/slider.lua
local _context = getContext()
local _input = _context:getInput()
local _painter = _context:getPainter()
local _style = _painter:getStyle()

exports('slider', function(min, value, max, w)
	local w = w or _context:getWidgetWidth()
	local h = _style.widget.height

	_context:beginDraw(w, h)

	local sliderStyle = _style.slider
	local newValue = value

	local isHovered = _input:isRectHovered(_painter:getX() - sliderStyle.tickMark.width / 2, _painter:getY(), w + sliderStyle.tickMark.width, h)
	if isHovered and (_input:isMouseDown() or _input:isMousePressed()) then
		newValue = math.min(max, math.max(min, min + ((_input:getMousePosX() - _painter:getX()) / w * (max + min))))
	end

	local sh = (h - sliderStyle.height) / 2

	_painter:setColor(_style.color.widget)
	_painter:move(0, sh)
	_painter:drawRect(w, sliderStyle.height)

	local sx = w * (returnValue or value) / (max + min)
	local tx = sx - sliderStyle.tickMark.width / 2
	local ty = -sliderStyle.tickMark.height / 4

	_painter:setColor(isHovered and _style.color.hover or _style.color.primary)
	_painter:move(tx, ty)
	_painter:drawRect(sliderStyle.tickMark.width, sliderStyle.tickMark.height)

	_context:endDraw()

	return newValue ~= value, newValue
end)

File: src/widgets/slider.ts
import { getCurrentContext } from '../../index';
import { numberEquals } from '../core/utils';
import { Color } from '../common/types';

interface SliderResult {
	isValueChanged: boolean;
	value: number;
}

export function declareExport() {
	globalThis.exports('slider', function (min: number, value: number, max: number, w?: number): SliderResult {
		const context = getCurrentContext();
		const input = context.getInput();
		const painter = context.getPainter();
		const style = painter.getStyle();

		w = (w ?? context.getWidgetWidth()) as number;
		const h = style.widget.height;

		context.beginDraw(w, h);

		const sliderStyle = style.slider;

		let newValue = value;

		const isHovered = input.isRectHovered(
			painter.getX() - sliderStyle.tickMark.width / 2,
			painter.getY(),
			w + sliderStyle.tickMark.width,
			h
		);

		if (isHovered && (input.getIsLmbDown() || input.getIsLmbPressed()))
			newValue = Math.min(max, Math.max(min, min + ((input.getMousePos().x - painter.getX()) / w) * (max + min)));

		const sh = (h - sliderStyle.height) / 2;

		const properties = style.getProperties(context.isWidgetHovered() ? 'slider:hover' : 'slider');

		painter.setColor(properties.get<Color>('background-color'));
		painter.move(0, sh);
		painter.drawRect(w, sliderStyle.height);

		const sx = (w * value) / (max + min);
		const tx = sx - sliderStyle.tickMark.width / 2;
		const ty = -sliderStyle.tickMark.height / 4;

		painter.setColor(properties.get<Color>('color'));
		painter.move(tx, ty);
		painter.drawRect(sliderStyle.tickMark.width, sliderStyle.tickMark.height);

		context.endDraw();

		return { isValueChanged: !numberEquals(newValue, value), value: newValue };
	});
}
